syntax = "proto3";

package ident.action_log;

import "google/protobuf/descriptor.proto";
import "google/protobuf/any.proto";

message ActionLog {
    uint64 id = 1;
    uint64 target_id = 2; // resource derived from action type
    
    ActionLogActor actor = 3;

    oneof action {
        ActionAclCreated acl_created = 4; 
        ActionAclUpdated acl_updated = 5;
        ActionAclDeleted acl_deleted = 6;

        ActionApplicationCreated application_created = 7;
        ActionApplicationAccessed application_accessed = 8;
        ActionApplicationDeleted application_deleted = 9;

        ActionGroupCreated group_created = 10;
        ActionGroupUpdated group_updated = 24;
        ActionGroupMemberAdded group_member_added = 11;
        ActionGroupMemberRemoved group_member_removed = 12;
        ActionGroupDeleted group_deleted = 13;

        ActionNetworkCreated network_created = 14;
        ActionNetworkUpdated network_updated = 15;
        ActionNetworkDeleted network_deleted = 16;

        ActionSessionCreated session_created = 17;

        ActionUserCreated user_created = 18;
        ActionUserSuspended user_suspended = 19;
        ActionUserPasswordChanged user_password_changed = 20;
        ActionUserDeleted user_deleted = 21;

        ActionWebhookCreated webhook_created = 22;
        ActionWebhookUpdated webhook_updated = 25;
        ActionWebhookDeleted webhook_deleted = 23;
        
        // (REMEMBER TO UPDATE) next idx = 26 
    }
}

message ActionLogActor {
    enum ActionLogActorType {
        ActionLogActorUser = 0;
        ActionLogActorSystem = 1;
        ActionLogActorAPIKey = 2; // TODO: necessary?
    }

    ActionLogActorType type = 1;
    uint64 id = 2;
}

message ActionLogChange {
    oneof change {
        ActionLogFieldChange field = 1;
        ActionLogKeyChange key = 2;
    }
}

message ActionLogFieldChange {
    google.protobuf.Any before = 1;
    google.protobuf.Any after = 2;
    string key = 3;
}

// map keys can be non-strings in protobuf, so we need Any here ...
message ActionLogKeyChange {
    google.protobuf.Any before = 1;
    google.protobuf.Any after = 2;
    string key = 3;
}

/* ACLs */
message ActionAclCreated {
    // all before should be None
    repeated ActionLogChange changes = 1;
}

message ActionAclUpdated {
    repeated ActionLogChange changes = 1;
}

message ActionAclDeleted {}


/* Applications */
message ActionApplicationCreated {
    // all before should be None
    repeated ActionLogChange changes = 1;
}

message ActionApplicationAccessed {
    oneof ip {
        uint32 ipv4 = 1;
        bytes ipv6 = 2;
    }
}


message ActionApplicationDeleted {}


/* Groups */
message ActionGroupCreated {
    repeated ActionLogChange changes = 1;
}

message ActionGroupUpdated {
    repeated ActionLogChange changes = 1;
}

message ActionGroupMemberAdded {
    uint64 member_id = 1;
}

message ActionGroupMemberRemoved {
    uint64 user_id = 1;
}

message ActionGroupDeleted {}


/* Networks */
message ActionNetworkCreated {
    repeated ActionLogChange changes = 1;
}

message ActionNetworkUpdated {
    repeated ActionLogChange changes = 1;
}

message ActionNetworkDeleted {}


/* Sessions */
message ActionSessionCreated {
    repeated ActionLogChange changes = 1;
}


/* Users */
message ActionUserCreated {
    repeated ActionLogChange changes = 1;
}

message ActionUserSuspended {
    string note = 1;
}

message ActionUserPasswordChanged {}

message ActionUserDeleted {}


/* Webhooks */
message ActionWebhookCreated {
    repeated ActionLogChange changes = 1;
}

message ActionWebhookUpdated {
    repeated ActionLogChange changes = 1;
}

message ActionWebhookDeleted {}



// used in other files to denote "hidden" fields in audit logging
extend google.protobuf.FieldOptions {
    optional bool ignored = 88681;
}